src/main/scala/mesosphere/marathon/core/volume/VolumesModule.scala

```diff
@@ -0,0 +1,94 @@
+package mesosphere.marathon.core.volume
+
+import com.wix.accord._
+import mesosphere.marathon.core.volume.providers._
+import mesosphere.marathon.state._
+
+/**
+  * VolumeProvider is an interface implemented by storage volume providers
+  */
+trait VolumeProvider[+T <: Volume] {
```

Comment:
Sometimes `sealed traits` + pattern matching provide an alternative that is similar to the visitor pattern. Maybe we can also use that.